Anni/wcicd (#16642)
cloudflare/cloudflare-docs: src/content/docs/workers/ci-cd/builds/build-image.mdx

---
pcx_content_type: reference
title: Build Image
description: Understand the build image used in Workers Builds.
sidebar:
  order: 3
---

Workers Builds uses a build image with support for a variety of languages and tools such as Node.js, Python, PHP, Ruby, and Go.

## Supported Languages and Tools

> In the following table, review the preinstalled versions for languages and tools included in the Cloudflare Workers' build image, and the environment variables and/or files available for [overriding the preinstalled version](/workers/ci-cd/builds/build-image/#overriding-default-versions):

Reviewer: ```suggestion
The following languages and tools are included in the Cloudflare Workers' build image. You can [override the preinstalled version](/workers/ci-cd/builds/build-image/#overriding-default-versions) by defining an environment variable or configuration file:
```

^ simpler language, suggest applying to all places where we start with "In the following table"

| Tool        | Default version  | Environment variable | File                         |
| ----------- | ---------------- | -------------------- | ---------------------------- |

> | **Go**      | 1.22.0           | `GO_VERSION`         |                              |

Reviewer 2: Update with:

`golang 1.23.0`
`ruby 3.3.5`
`bun 1.1.28`
`hugo 0.134.3`

| **Node.js** | 22               | `NODE_VERSION`       | .nvmrc, .node-version        |
| **Python**  | 3.12.5           | `PYTHON_VERSION`     | .python-version, runtime.txt |
| **Ruby**    | 3.3.4            | `RUBY_VERSION`       | .ruby-version                |
| **Bun**     | 1.1.22           | `BUN_VERSION`        |                              |
| **Hugo**    | extended_0.130.0 | `HUGO_VERSION`       |                              |

> The default versions will be updated regularly to the latest minor version. No major version updates will be made without notice. If you need a specific minor version, please specify it by overriding the default version.

Reviewer: ```suggestion
Cloudflare updates the default versions regularly to the latest minor version. We do not make major version updates to preinstalled versions without advance notice. If you need a specific version, you should specify it by overriding the default version.
```

> The default versions will be updated regularly to the latest minor version. No major version updates will be made without notice. If you need a specific minor version, please specify it by overriding the default version.

Reviewer: Dupe

> ### Overriding Default Versions

Reviewer: ```suggestion
### How to Override Preinstalled Versions
```

We should be consistent — either call these `preinstalled` or `default`, not mix and match

> If you need to override a [specific version](/workers/ci-cd/builds/build-image/#overriding-default-versions) of a language or tool within the image, you can specify it as a [build environment variable](/workers/ci-cd/builds/build-configuration/#build-settings), or set the relevant file in your source code as shown above.

Reviewer: We already say this above, and I think this is linking to itself right now?

To set the version using a build environment variables, you can:

1. Find the environment variable name for the language or tool and desired version (e.g. `NODE_VERSION = 22`)

> 2. Add and save the environment variable on the dashboard by going to **Settings** > **Build** > **Build Variables and Secrets* in your Workers project

Reviewer: This is another instance where the reader will (rightly) ask — can I set these in my own codebase? In Terraform? etc.

When something isn't yet supported, a ton of the value that docs can provide can be to just be explicit and clear about it upfront, so that someone doesn't have to chase this down themselves. Turn frustration into simple feature request, etc.

Or, to set the version by adding a file to your project, you can:

1. Find the filename for the language or tool (e.g. `.nvmrc`)
2. Add the specified file name to the root directory and add the desired version number as the contents of the file (e.g. 22)

## Pre-installed Packages

> In the following table, review the pre-installed packages in the build image. The packages are installed with `apt`, a package manager for Linux distributions.

Reviewer: Is there anything important to say here about versions?

---
pcx_content_type: reference
title: Build Image
description: Understand the build image used in Workers Builds.
---

Workers Builds' build environment has broad support for a variety of languages, such as Ruby, Node.js, Python, PHP, and Go.

> 

Reviewer 3: Very nit: I don't like the possessive Builds: Workers Builds'

> If you need to use a [specific version](/workers/ci-cd/builds/build-image/#overriding-default-versions) of a language, (for example, Node.js or Ruby) you can specify it by providing an associated environment variable in your build configuration, or setting the relevant file in your source code.

Reviewer 3: This is linked in the paragraph below, you may want to just remove this link?

| **Ruby**    | 3.3.4            | Any version        | `RUBY_VERSION`       | .ruby-version                |
| **Bun**     | 1.1.22           | Any version        | `BUN_VERSION`        |                              |
| **Hugo**    | extended_0.130.0 | Any version        | `HUGO_VERSION`       |                              |

:::note[Any version]
Under Supported versions, "Any version" refers to support for all versions of the language or tool including versions newer than the Default version.
:::

### Frameworks

To use a specific version of a framework, specify it in the project's package manager configuration file.
For example, if you use Gatsby, your `package.json` should include the following:

```
"dependencies": {
	"gatsby": "^5.13.7",
}
```

When your build starts, version 5.13.7 of Gatsby will be installed using `npm install`.

## Overriding default versions

To override default versions of languages and tools in the build system, you can either set the desired version through environment variables or by adding files to your project.

To set the version using **environment variables**, you can:

> 1. Find the environment variable name for the language or tool in [this table]().

Reviewer 3: link

2. Add the environment variable on the dashboard by going to **Settings** > **Environmnet variables** in your Workers project, or [add the environment variable via Wrangler](/workers/configuration/environment-variables/#add-environment-variables-via-wrangler).

Or, to set the version by **adding a file** to your project, you can:

1. Find the environment variable name for the language or tool in [this table]().

> 2. Add the specified file name to the root directory of your project, and add the desired version number as the contents of the file.

Reviewer 4: ```suggestion
2. Add the specified file name to the root directory of your project and add the desired version number as the contents of the file.
```
